pyogrio/tests/test_geopandas_io.py

```diff
@@ -207,6 +218,144 @@ def test_read_fids_force_2d(test_fgdb_vsi):
+    if naturalearth_lowres_all_ext.suffix == ".gpkg":
+        with pytest.raises(Exception, match="In ExecuteSQL().*"):
+            read_dataframe(naturalearth_lowres_all_ext, sql="invalid")
+    else:
+        with pytest.raises(Exception, match="SQL Expression Parsing Error"):
+            read_dataframe(naturalearth_lowres_all_ext, sql="invalid")
+
+    with pytest.raises(
+            ValueError, match="'sql' paramater cannot be combined with 'layer'"):
+        read_dataframe(naturalearth_lowres_all_ext, sql="whatever", layer="invalid")
+
+
+def test_read_sql_columns_where(naturalearth_lowres_all_ext):
+    sql = "SELECT iso_a3 AS iso_a3_renamed, name, pop_est FROM naturalearth_lowres"
+    df = read_dataframe(
+        naturalearth_lowres_all_ext,
+        sql=sql,
+        sql_dialect="OGRSQL",
+        columns=["iso_a3_renamed", "name"],
+        where="iso_a3_renamed IN ('CAN', 'USA', 'MEX')",
+    )
+    assert len(df.columns) == 3
+    assert len(df) == 3
+    assert df.iso_a3_renamed.tolist() == ["CAN", "USA", "MEX"]
+
+
+def test_read_sql_columns_where_bbox(naturalearth_lowres_all_ext):
+    sql = "SELECT iso_a3 AS iso_a3_renamed, name, pop_est FROM naturalearth_lowres"
+    df = read_dataframe(
+        naturalearth_lowres_all_ext,
+        sql=sql,
+        sql_dialect="OGRSQL",
+        columns=["iso_a3_renamed", "name"],
+        where="iso_a3_renamed IN ('CAN', 'USA', 'MEX')",
+        bbox=(-140, 20, -100, 40)
+    )
+    assert len(df.columns) == 3
+    assert len(df) == 2
+    assert df.iso_a3_renamed.tolist() == ["USA", "MEX"]
+
+
+def test_read_sql_skip_max(naturalearth_lowres_all_ext):
+    sql = """SELECT *
+               FROM naturalearth_lowres
+              WHERE iso_a3 IN ('CAN', 'MEX', 'USA')
+              ORDER BY name"""
+    df = read_dataframe(
+            naturalearth_lowres_all_ext, sql=sql, skip_features=1,
+            max_features=1, sql_dialect="OGRSQL")
+    assert len(df.columns) == 6
+    assert len(df) == 1
+    assert df.iso_a3.tolist() == ["MEX"]
+
+    sql = "SELECT * FROM naturalearth_lowres LIMIT 1"
+    df = read_dataframe(
+            naturalearth_lowres_all_ext, sql=sql, max_features=3, sql_dialect="OGRSQL")
+    assert len(df) == 1
+
+    sql = "SELECT * FROM naturalearth_lowres LIMIT 1"
+    with pytest.raises(ValueError, match="'skip_features' must be between 0 and 0"):
+        _ = read_dataframe(
+                naturalearth_lowres_all_ext, sql=sql, skip_features=1, 
+                sql_dialect="OGRSQL")
+
+
+def test_read_sql_dialect_sqlite(naturalearth_lowres_all_ext):
+    # Should return singular item
+    sql = "SELECT * FROM naturalearth_lowres WHERE iso_a3 = 'CAN'"
+    df = read_dataframe(naturalearth_lowres_all_ext, sql=sql, sql_dialect="SQLITE")
+    assert len(df) == 1
+    assert len(df.columns) == 6
+    assert df.iloc[0].iso_a3 == "CAN"
+    area_canada = df.iloc[0].geometry.area
+
+    # Use spatialite function
```

Comment:
Suggest splitting this into a separate test that can be executed or skipped in whole, and then limit that test to GPKG inputs.

Reply:
I now remembered a way to get this working anyway. GDAL is actually always build with spatialite support as far as I know, but for some odd reason spatialite is not activated in sqlite on those images. So as long as you don't run such queries on non-native-sqlite files, the spatialite used will be the one "inside" GDAL. If you run such a query on a GPKG, by default GDAL will just pass the sql through to sqlite... and if spatialite isn't activated there, it doesn't work.

However, you can force gdal not to pass through the SQLITE query by specifying "INDIRECT_SQLITE" as sql_dialect, and then it works, even though performance might be impacted.

I'll change the documentation accordingly, but regarding the tests: do you still like the tests separated in this case? When it is detected that spatialite is not available within sqlite, the won't be skipped anymore but the sqlite_dialect will be changed to INDIRECT_SQLITE.

Reply:
> do you still like the tests separated in this case

Yes, I think smaller focused test cases are better.

> When it is detected that spatialite is not available within sqlite, the won't be skipped anymore but the sqlite_dialect will be changed to INDIRECT_SQLITE

This is a good solution, thank you for the additional investigation.

> request for the gdal-ubuntu-small image that spatialite would be activated within sqlite as well

Certainly could be raised as an issue in the GDAL repo; I'm not sure of the implications for image size / complexity / license (could well be the latter).

Reply:
OK, I kept it in a separate test...